util.go

```diff
@@ -50,19 +53,180 @@ func parseTag(path string) string {
+			if parts[0] == "require" {
+				depMap, err = processRequiresSection(s)
+				if err != nil {
+					return nil, err
+				}
+			} else if parts[0] == "replace" {
+				replaceMap, err = processReplaceSection(s)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
+	if err := s.Err(); err != nil {
+		return nil, err
+	}
+	for depName, version := range replaceMap {
+		if oldDep, ok := depMap[depName]; ok {
+			oldDep.Commit = version
+		} else {
+			logrus.Debugf("dependency %s found in replace section, but doesn't exist in requires section. Skipping", depName)
+			continue
+		}
+	}
+	var deps []dependency
+	for _, dep := range depMap {
+		deps = append(deps, *dep)
+	}
+
+	return deps, nil
+}
+
+func processReplaceSection(s *bufio.Scanner) (map[string]string, error) {
+	replaceMap := make(map[string]string)
+	for s.Scan() {
+		ln := sanitizeLine(s.Text(), "//")
+		if ln == "" {
+			continue
+		}
+		parts := strings.Fields(ln)
+		numParts := len(parts)
+		if numParts != 4 {
+			if numParts == 1 && parts[0] == ")" {
+				// this is the end of the requires section, break out to process the others
+				break
+			}
+			return nil, fmt.Errorf("invalid config format: %s", ln)
+		}
+
+		commitOrVersion := getCommitOrVersion(parts[3])
+		if commitOrVersion == "" {
+			return nil, fmt.Errorf("invalid go.mod file, poorly formatted version in replace section %s", parts[3])
+		}
+		replaceMap[parts[0]] = commitOrVersion
+	}
+	if err := s.Err(); err != nil {
+		return nil, err
+	}
+	return replaceMap, nil
+}
+
+func processRequiresSection(s *bufio.Scanner) (map[string]*dependency, error) {
+	depMap := make(map[string]*dependency)
+	for s.Scan() {
+		ln := sanitizeLine(s.Text(), "//")
+		if ln == "" {
+			continue
+		}
+		parts := strings.Fields(ln)
+		numParts := len(parts)
+
+		if numParts != 2 {
+			if numParts == 1 && parts[0] == ")" {
+				// this is the end of the requires section, break out to process the others
+				break
+			}
+			return nil, fmt.Errorf("invalid config format: %s", ln)
+		}
+
+		commitOrVersion := getCommitOrVersion(parts[1])
+		if commitOrVersion == "" {
+			return nil, fmt.Errorf("invalid go.mod file, poorly formatted version in require section %s", parts[1])
+		}
+
+		depMap[parts[0]] = &dependency{
+			Name:     parts[0],
+			Commit:   commitOrVersion,
+			CloneURL: "git://" + parts[0],
+		}
+	}
+	if err := s.Err(); err != nil {
+		return nil, err
+	}
+	return depMap, nil
+}
+
+func sanitizeLine(line, commentDelim string) string {
+	ln := strings.TrimSpace(line)
+	if ln == "" {
+		return ""
+	}
+	cidx := strings.Index(ln, commentDelim)
+	// whole line is commented
+	if cidx == 0 {
+		return ""
+	}
+	if cidx > 0 {
+		ln = ln[:cidx]
+	}
+
+	return strings.TrimSpace(ln)
+}
+
+func getCommitOrVersion(cov string) string {
+	// parse the commit or version. It'll either be of the form
+	// v0.0.0 or v0.0.0-date-commitID. Split by '-' to check
+	dashFields := strings.FieldsFunc(cov, func(c rune) bool { return c == '-' })
```

Comment:
Do they have a published regex for this?

I don't really understand this madness, but just using the 3rd parse seems mostly right
`v0.8.0-dev.2.0.20190604151032-3c26b4e7495e`
`v2.7.1-0.20190205005809-0d3efadf0154+incompatible`
`v0.0.0-20170609045927-2bca23e0e452`
`v1.14.0-0.20190319215453-e7b5f7dbe98c`

Reply:
I don't know of one, I formulated this based off of anecdotal examples. I agree, we should use the 3rd part, if it exists.